## Load Packages and Data

In [1]:
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import pandas as pd
from sklearn.ensemble import GradientBoostingRegressor, GradientBoostingClassifier
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.model_selection import KFold
from sklearn.metrics import mean_squared_error, r2_score


# this extension properly formats a cell after it is run
# !{sys.executable} -m pip install nb_black # UNCOMMENT TO INSTALL
%load_ext nb_black
%matplotlib inline

# Set the maximum number of rows to 200
pd.set_option("display.max_rows", 200)

# Set the maximum number of columns to 200
pd.set_option("display.max_columns", 200)

<IPython.core.display.Javascript object>

In [2]:
# Load Data
train_set = pd.read_csv("../data/train_set.csv")
test_set = pd.read_csv("../data/test_set.csv")

<IPython.core.display.Javascript object>

In [3]:
train_set.columns

Index(['PID', 'GrLivArea', 'SalePrice', 'MSSubClass', 'MSZoning',
       'LotFrontage', 'LotArea', 'Street', 'Alley', 'LotShape', 'LandContour',
       'Utilities', 'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1',
       'Condition2', 'BldgType', 'HouseStyle', 'OverallQual', 'OverallCond',
       'YearBuilt', 'YearRemodAdd', 'RoofStyle', 'RoofMatl', 'Exterior1st',
       'Exterior2nd', 'MasVnrType', 'MasVnrArea', 'ExterQual', 'ExterCond',
       'Foundation', 'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1',
       'BsmtFinSF1', 'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF',
       'Heating', 'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF',
       '2ndFlrSF', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath',
       'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual', 'TotRmsAbvGrd',
       'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType', 'GarageYrBlt',
       'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual', 'GarageCond',
       'PavedDrive',

<IPython.core.display.Javascript object>

## Function from Creating_test_train notbook with a few changes ##

In [4]:
## Group Streets by Sale Price ##

# creating a function to create groups based on mean sale price of streets in each
# neighborhood of the training df and adds a new column with the groups created
# It then uses the groups created using the training dataframe and creates a new column
# in the testing dataframe by mapping these groups.  This will result in missing values
# If there is a street neighborhood combination that was in the training dataset but 
# not the testing dataset


def group_neighbor_streets_by_saleprice(
    traindf=train_set, testdf=test_set, num_quantiles=10  # notice the difference in this line!!
):
    # Calculate the mean sale price for each street in the training df
    street_prices = traindf.groupby("Neighborhood_st")["SalePrice"].mean()
    
    # Group the streets into the specified number of quantiles based on sale price
    labels = [f"group_{i+1}" for i in range(num_quantiles)]
    groups = pd.qcut(street_prices, q=num_quantiles, labels=range(1, num_quantiles + 1))
    
    # Create a dictionary that maps each street name to its corresponding sale price group label
    street_group_dict = dict(zip(street_prices.index, groups))
    
    # Add a new column to the training dataframe with the street price groups
    traindf["StreetPriceGroup"] = traindf["Neighborhood_st"].map(street_group_dict)
    
    # Add a new column to the testing dataframe with the street price groups
    testdf["StreetPriceGroup"] = testdf["Neighborhood_st"].map(street_group_dict)
    return street_group_dict


<IPython.core.display.Javascript object>

In [5]:
# the num of quantiles can be changed and it is assigned to d which is the dictionary that
# will be used to fill in the missing values

d = group_neighbor_streets_by_saleprice(
    traindf=train_set,  # the difference here as well!!
    testdf=test_set,  # the difference here as well!!
    num_quantiles=10,
)

<IPython.core.display.Javascript object>

In [6]:
## Supporting Function to group_neighbor_streets_by_saleprice to FILL_NA's ##

# Dictionary stores missing values in the test df with another group in the same neighborhood


def fill_na(testdf=test_set, d={}):  # DIFFERENT HERE TOO
    # Extract the first part of the string in the "Neighborhood_st" column
    testdf["Neighborhood_prefix"] = testdf["Neighborhood_st"].map(
        lambda x: x.split("_")[0]
    )

    # Create a new dict that only contains the neighborhood
    new_dict = {k.split("_")[0]: v for k, v in d.items()}

    # Create a list of PIDs with missing StreetPriceGroup values
    na_pid_list = testdf[testdf["StreetPriceGroup"].isna()]["PID"].tolist()

    # Create a Boolean mask to filter the DataFrame
    mask = testdf["PID"].isin(na_pid_list)

    # Apply the dictionary mapping only to the filtered rows
    testdf.loc[mask, "StreetPriceGroup"] = testdf[mask]["Neighborhood_prefix"].map(
        new_dict
    )

    # Drop the column since there is no more use for it
    testdf.drop("Neighborhood_prefix", axis=1, inplace=True)

<IPython.core.display.Javascript object>

In [7]:
fill_na(test_set, d)

<IPython.core.display.Javascript object>

In [8]:
# Remove the column that was used to create groupings
train_set.drop("Neighborhood_st", axis=1, inplace=True)
test_set.drop("Neighborhood_st", axis=1, inplace=True)

<IPython.core.display.Javascript object>

In [9]:
# Looks like PID should be removed and SalePrice moved to the first column
train_set

PID  GrLivArea  SalePrice  MSSubClass MSZoning  LotFrontage  \
0     534477270       1661     165500          80       RL         80.0   
1     903233140        854     132000          45       RM         51.0   
2     923228370       1092      85500         160       RM         21.0   
3     904101070       3086     200500          50       RL        138.0   
4     533236040       1441     174000         160       FV         24.0   
...         ...        ...        ...         ...      ...          ...   
2041  904100190        704      88750          20       RL         50.0   
2042  534128190       2640     256900          60       RL         80.0   
2043  528142110       1944     360000          20       RL         90.0   
2044  534127140       1082     162500          85       RL          0.0   
2045  909428240       1906     224000          20       RL        100.0   

      LotArea Street     Alley LotShape LandContour Utilities LotConfig  \
0        9600   Pave  No_Alley      Reg         Lvl    AllPub    Inside   
1        6120   Pave  No_Alley      Reg         Lvl    AllPub    Corner   
2        1890   Pave  No_Alley      Reg         Lvl    AllPub    Inside   
3       18030   Pave  No_Alley      IR1         Bnk    AllPub    Inside   
4        2645   Pave      Pave      Reg         Lvl    AllPub    Inside   
...       ...    ...       ...      ...         ...       ...       ...   
2041     4280   Pave  No_Alley      IR1         Lvl    AllPub    Inside   
2042    10400   Pave  No_Alley      Reg         Lvl    AllPub    Inside   
2043    12378   Pave  No_Alley      IR1         Lvl    AllPub    Inside   
2044     8723   Pave  No_Alley      IR1         Lvl    AllPub    Inside   
2045    17500   Pave  No_Alley      Reg         Bnk    AllPub    Corner   

     LandSlope Neighborhood Condition1 Condition2 BldgType HouseStyle  \
0          Gtl        NAmes       Norm       Norm     1Fam       SLvl   
1          Gtl      BrkSide       Norm       Norm     1Fam     1.5Unf   
2          Gtl      MeadowV       Norm       Norm   TwnhsE     2Story   
3          Gtl      ClearCr       Norm       Norm     1Fam     1.5Fin   
4          Gtl      Somerst       Norm       Norm    Twnhs     2Story   
...        ...          ...        ...        ...      ...        ...   
2041       Gtl       Sawyer       Norm       Norm     1Fam     1Story   
2042       Gtl       NWAmes       PosA       Norm     1Fam     2Story   
2043       Gtl      NridgHt       Norm       Norm     1Fam     1Story   
2044       Gtl       NWAmes       PosN       Norm     1Fam     SFoyer   
2045       Mod      Crawfor       PosA       Norm     1Fam     1Story   

      OverallQual  OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl  \
0               6            6       1955          1996       Hip  CompShg   
1               7            8       1929          2001     Gable  CompShg   
2               4            6       1972          1972     Gable  CompShg   
3               5            6       1946          1994     Gable  CompShg   
4               8            5       1999          2000     Gable  CompShg   
...           ...          ...        ...           ...       ...      ...   
2041            4            9       1946          2001     Gable  CompShg   
2042            6            7       1967          1997     Gable  CompShg   
2043            9            5       2003          2004     Gable  CompShg   
2044            6            6       1969          1969     Gable  CompShg   
2045            6            5       1954          1954     Gable  CompShg   

     Exterior1st Exterior2nd MasVnrType  MasVnrArea  ExterQual  ExterCond  \
0        AsbShng     AsbShng       None         0.0          2          2   
1        Wd Sdng     Wd Sdng       None         0.0          2          2   
2        CemntBd     CmentBd       None         0.0          2          2   
3        MetalSd     MetalSd       None         0.0          2          2   
4        Me

<IPython.core.display.Javascript object>

In [10]:
# Remove PID and move SalePrice to first column
# I need to do this in the future and save the csv's after so we dont have to do this each time

# TRAIN
train_set = train_set.iloc[
    :,
    train_set.columns.tolist().index("SalePrice") : (
        train_set.columns.tolist().index("SalePrice") + 1
    ),
].join(train_set.drop(columns=["SalePrice", "PID"]))

# TEST
test_set = test_set.iloc[
    :,
    test_set.columns.tolist().index("SalePrice") : (
        test_set.columns.tolist().index("SalePrice") + 1
    ),
].join(test_set.drop(columns=["SalePrice", "PID"]))

<IPython.core.display.Javascript object>

## One Hot Encoding for Categorical Values

In [11]:
# In order to do regression, I need to turn categorical values into numerical values
train_set.dtypes

SalePrice             int64
GrLivArea             int64
MSSubClass            int64
MSZoning             object
LotFrontage         float64
LotArea               int64
Street               object
Alley                object
LotShape             object
LandContour          object
Utilities            object
LotConfig            object
LandSlope            object
Neighborhood         object
Condition1           object
Condition2           object
BldgType             object
HouseStyle           object
OverallQual           int64
OverallCond           int64
YearBuilt             int64
YearRemodAdd          int64
RoofStyle            object
RoofMatl             object
Exterior1st          object
Exterior2nd          object
MasVnrType           object
MasVnrArea          float64
ExterQual             int64
ExterCond             int64
Foundation           object
BsmtQual              int64
BsmtCond              int64
BsmtExposure          int64
BsmtFinType1          int64
BsmtFinSF1          

<IPython.core.display.Javascript object>

In [12]:
## DUMMIFY categorical columns into numerical values 1,2,3,4 etc. ##


# List categorical columns to be dummified for train/test
cat_cols = train_set.select_dtypes(include=['object']).columns.tolist()

# Dummify the categorical columns in both train/test set
train_dummies = pd.get_dummies(train_set, columns=cat_cols)
test_dummies = pd.get_dummies(test_set, columns=cat_cols)

# Ensure the same dummy columns are present in both train and test sets
train_dummies, test_dummies = train_dummies.align(test_dummies, join='left', axis=1)

# Fill any missing columns in the test set with zeros
test_dummies.fillna(0, inplace=True)


<IPython.core.display.Javascript object>

In [16]:
# Did it work? yes!
train_dummies

SalePrice  GrLivArea  MSSubClass  LotFrontage  LotArea  OverallQual  \
0        165500       1661          80         80.0     9600            6   
1        132000        854          45         51.0     6120            7   
2         85500       1092         160         21.0     1890            4   
3        200500       3086          50        138.0    18030            5   
4        174000       1441         160         24.0     2645            8   
...         ...        ...         ...          ...      ...          ...   
2041      88750        704          20         50.0     4280            4   
2042     256900       2640          60         80.0    10400            6   
2043     360000       1944          20         90.0    12378            9   
2044     162500       1082          85          0.0     8723            6   
2045     224000       1906          20        100.0    17500            6   

      OverallCond  YearBuilt  YearRemodAdd  MasVnrArea  ExterQual  ExterCond  \
0               6       1955          1996         0.0          2          2   
1               8       1929          2001         0.0          2          2   
2               6       1972          1972         0.0          2          2   
3               6       1946          1994         0.0          2          2   
4               5       1999          2000       456.0          3          2   
...           ...        ...           ...         ...        ...        ...   
2041            9       1946          2001         0.0          2          3   
2042            7       1967          1997       256.0          2          2   
2043            5       2003          2004         0.0          3          2   
2044            6       1969          1969         0.0          2          2   
2045            5       1954          1954       420.0          2          2   

      BsmtQual  BsmtCond  BsmtExposure  BsmtFinType1  BsmtFinSF1  \
0            3         3             4             4       831.0   
1            3         3             1             1         0.0   
2            3         3             1             3       294.0   
3            3         3             1             3       152.0   
4            4         3             1             1         0.0   
...        ...       ...           ...           ...         ...   
2041         2         3             1             1         0.0   
2042         3         3             1             1         0.0   
2043         5         3             3             6      1274.0   
2044         4         3             4             4       973.0   
2045         3         3             4             2       784.0   

      BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  HeatingQC  1stFlrSF  \
0                1         0.0      161.0        992.0          3      1661   
1                1         0.0      832.0        832.0          4       854   
2                1         0.0      252.0        546.0          2       546   
3                4       469.0      977.0       1598.0          2      1636   
4                1         0.0      776.0        776.0          4       764   
...            ...         ...        ...          ...        ...       ...   
2041             1         0.0      560.0        560.0          4       704   
2042             1         0.0      932.0        932.0          3      1271   
2043             1         0.0      622.0       1896.0          4      1944   
2044             1         0.0        0.0        973.0          4      1082   
2045             4       435.0       91.0       1310.0          4      1906   

      2ndFlrSF  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  \
0            0           1.0           0.0         1         0             3   
1            0           0.0           0.0         1         0             2   
2          546           0.0           0.0         1         1             3   
3          971           0.0      

<IPython.core.display.Javascript object>

In [20]:
# Lot's of Columns now that we one-hot encoded!
print(train_dummies.shape)
print(test_dummies.shape)

(2046, 266)
(512, 266)


<IPython.core.display.Javascript object>

In [18]:
# Triple Check that there are no object values
train_strings = train_dummies.select_dtypes(include=["object"]).columns
test_strings = test_dummies.select_dtypes(include=["object"]).columns
print(train_strings)
print(test_strings)

Index([], dtype='object')
Index([], dtype='object')


<IPython.core.display.Javascript object>

## Gradient Boosting Regression Model

In [21]:
gbr = GradientBoostingRegressor()
kf = KFold(n_splits=5, shuffle=True, random_state=42)
X = train_dummies.loc[:, train_dummies.columns != "SalePrice"]
y = train_dummies.loc[:, "SalePrice"]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)


<IPython.core.display.Javascript object>

In [22]:
# FIT
gbr.fit(X, y)

GradientBoostingRegressor()

<IPython.core.display.Javascript object>

In [23]:
# k-fold cross-validation

# Create the Gradient Boosted Regression model and evaluate it using k-fold cross-validation
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.metrics import mean_squared_error, r2_score

mse_scores = []
r2_scores = []

for train_index, test_index in kf.split(X):
    X_train, X_test = X.iloc[train_index], X.iloc[test_index]
    y_train, y_test = y.iloc[train_index], y.iloc[test_index]

    model = GradientBoostingRegressor(
        n_estimators=100, learning_rate=0.1, max_depth=3, random_state=42
    )
    model.fit(X_train, y_train)

    y_pred = model.predict(X_test)
    mse = mean_squared_error(y_test, y_pred)
    r2 = r2_score(y_test, y_pred)

    mse_scores.append(mse)
    r2_scores.append(r2)

print("MSE scores:", mse_scores)
print("R-squared scores:", r2_scores)

MSE scores: [322746634.2791484, 465856217.6223381, 369490124.33649755, 949124370.7377828, 325101453.40383786]
R-squared scores: [0.9352264122721555, 0.921725186226128, 0.9312188870117949, 0.8501043521734962, 0.9400063021503235]


<IPython.core.display.Javascript object>

The output shows the MSE and R-squared scores for each fold of the cross-validation. 
The MSE scores range from 322,746,634 to 949,124,370 while the R-squared scores range from 0.8501 to 0.9400. 

These scores indicate how well the model is performing on the testing data, with higher R-squared scores and lower MSE scores indicating better performance. 

The fact that the R-squared scores are relatively high and the MSE scores are relatively low suggests that the model is performing well on the data.